Cannot use the package with strict typescript config

## Short description

Currently this package cannot be used when we use strict typescript configuration.

## Steps to reproduce

1. `nest new foo --strict` (Note the `--strict` argument)
2. `npm install --save @itgorillaz/configify`
3. Add `ConfigifyModule` in the imports array in `app.module.ts`:

```app.module.ts
@Module({
  imports: [ConfigifyModule.forRootAsync()],
  controllers: [AppController],
  providers: [AppService],
})
export class AppModule {}
```

4. `npm run start:dev`

We get the following error:

```
node_modules/@itgorillaz/configify/src/configify.module.ts:235:15 - error TS2769: No overload matches this call.
  Overload 1 of 2, '(...items: ConcatArray<never>[]): never[]', gave the following error.
    Argument of type 'string | string[] | undefined' is not assignable to parameter of type 'ConcatArray<never>'.
      Type 'undefined' is not assignable to type 'ConcatArray<never>'.
  Overload 2 of 2, '(...items: ConcatArray<never>[]): never[]', gave the following error.
    Argument of type 'string | string[] | undefined' is not assignable to parameter of type 'ConcatArray<never>'.
      Type 'undefined' is not assignable to type 'ConcatArray<never>'.

235       .concat(path, this.DEFAULT_CONFIG_FILES)
                  ~~~~


node_modules/@itgorillaz/configify/src/configuration/configuration.registry.ts:22:24 - error TS2345: Argument of type 'any' is not assignable to parameter of type 'never'.

22     this.registry.push(type);
                          ~~~~

node_modules/@itgorillaz/configify/src/configuration/parsers/configuration-parser.factory.ts:30:12 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{ env: DotEnvConfigurationParser; yml: YamlConfigurationParser; yaml: YamlConfigurationParser; json: JsonConfigurationParser; }'.
  No index signature with a parameter of type 'string' was found on type '{ env: DotEnvConfigurationParser; yml: YamlConfigurationParser; yaml: YamlConfigurationParser; json: JsonConfigurationParser; }'.

30     return this.parsers[ext];
              ~~~~~~~~~~~~~~~~~

node_modules/@itgorillaz/configify/src/configuration/parsers/configuration-parser.factory.ts:51:5 - error TS2322: Type 'string | undefined' is not assignable to type 'string'.
  Type 'undefined' is not assignable to type 'string'.

51     return file.split('.').pop();
       ~~~~~~

node_modules/@itgorillaz/configify/src/configuration/parsers/yaml-configuration.parser.ts:2:23 - error TS7016: Could not find a declaration file for module 'js-yaml'. '/home/safal/work/templates/nestjs/template/node_modules/js-yaml/index.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/js-yaml` if it exists or add a new declaration (.d.ts) file containing `declare module 'js-yaml';`

2 import * as yaml from 'js-yaml';
                        ~~~~~~~~~

node_modules/@itgorillaz/configify/src/configuration/resolvers/aws/parameter-store-configuration.resolver.ts:39:46 - error TS18048: 'e.error' is possibly 'undefined'.

39           .map((e) => `${e.key}: ${e.id} - ${e.error.message}`)
                                                ~~~~~~~

node_modules/@itgorillaz/configify/src/configuration/resolvers/aws/secrets-manager-configuration.resolver.ts:42:46 - error TS18048: 'e.error' is possibly 'undefined'.

42           .map((e) => `${e.key}: ${e.id} - ${e.error.message}`)
                                                ~~~~~~~

node_modules/@itgorillaz/configify/src/interpolation/variables.ts:44:7 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{}'.
  No index signature with a parameter of type 'string' was found on type '{}'.

44       expanded[key] =
         ~~~~~~~~~~~~~
```

## System information

- OS: Arch Linux x86_64
- Kernel: 6.10.7-arch1-1
- Node.js version: v22.7.0
- Package manager used in this project: npm
- NPM version: 10.8.3

`package.json` contents:

```json
{
  "name": "foo",
  "version": "0.0.1",
  "description": "",
  "author": "",
  "private": true,
  "license": "UNLICENSED",
  "scripts": {
    "build": "nest build",
    "format": "prettier --write \"src/**/*.ts\" \"test/**/*.ts\"",
    "start": "nest start",
    "start:dev": "nest start --watch",
    "start:debug": "nest start --debug --watch",
    "start:prod": "node dist/main",
    "lint": "eslint \"{src,apps,libs,test}/**/*.ts\" --fix",
    "test": "jest",
    "test:watch": "jest --watch",
    "test:cov": "jest --coverage",
    "test:debug": "node --inspect-brk -r tsconfig-paths/register -r ts-node/register node_modules/.bin/jest --runInBand",
    "test:e2e": "jest --config ./test/jest-e2e.json"
  },
  "dependencies": {
    "@itgorillaz/configify": "^1.2.3",
    "@nestjs/common": "^10.0.0",
    "@nestjs/core": "^10.0.0",
    "@nestjs/platform-express": "^10.0.0",
    "reflect-metadata": "^0.2.0",
    "rxjs": "^7.8.1"
  },
  "devDependencies": {
    "@nestjs/cli": "^10.0.0",
    "@nestjs/schematics": "^10.0.0",
    "@nestjs/testing": "^10.0.0",
    "@types/express": "^4.17.17",
    "@types/jest": "^29.5.2",
    "@types/node": "^20.3.1",
    "@types/supertest": "^6.0.0",
    "@typescript-eslint/eslint-plugin": "^8.0.0",
    "@typescript-eslint/parser": "^8.0.0",
    "eslint": "^8.42.0",
    "eslint-config-prettier": "^9.0.0",
    "eslint-plugin-prettier": "^5.0.0",
    "jest": "^29.5.0",
    "prettier": "^3.0.0",
    "source-map-support": "^0.5.21",
    "supertest": "^7.0.0",
    "ts-jest": "^29.1.0",
    "ts-loader": "^9.4.3",
    "ts-node": "^10.9.1",
    "tsconfig-paths": "^4.2.0",
    "typescript": "^5.1.3"
  },
  "jest": {
    "moduleFileExtensions": [
      "js",
      "json",
      "ts"
    ],
    "rootDir": "src",
    "testRegex": ".*\\.spec\\.ts$",
    "transform": {
      "^.+\\.(t|j)s$": "ts-jest"
    },
    "collectCoverageFrom": [
      "**/*.(t|j)s"
    ],
    "coverageDirectory": "../coverage",
    "testEnvironment": "node"
  }
}
```

--- a/src/configify.module.ts
+++ b/src/configify.module.ts
@@ -231,11 +231,20 @@ export class ConfigifyModule {
    * @returns {string[]}                list of configuration files
    */
   private static resolveConfigurationFiles(path?: string | string[]): string[] {
-    return []
-      .concat(path, this.DEFAULT_CONFIG_FILES)
+    const res: string[] = [];
+    res
+      .concat(
+        path !== undefined
+          ? typeof path == 'string'
+            ? [path]
+            : [...path]
+          : [],
+        this.DEFAULT_CONFIG_FILES,
+      )
       .filter(
         (file) =>
           fs.existsSync(file) && ConfigurationParserFactory.supports(file),
       );
+    return res;
   }
 }

--- a/src/configify.module.ts
+++ b/src/configify.module.ts
@@ -231,8 +231,7 @@ export class ConfigifyModule {
    * @returns {string[]}                list of configuration files
    */
   private static resolveConfigurationFiles(path?: string | string[]): string[] {
-    const res: string[] = [];
-    res
+    return ([] as string[])
       .concat(
         path !== undefined
           ? typeof path == 'string'
@@ -245,6 +244,5 @@ export class ConfigifyModule {
         (file) =>
           fs.existsSync(file) && ConfigurationParserFactory.supports(file),
       );
-    return res;
   }
 }

--- a/src/configify.module.ts
+++ b/src/configify.module.ts
@@ -232,14 +232,7 @@ export class ConfigifyModule {
    */
   private static resolveConfigurationFiles(path?: string | string[]): string[] {
     return ([] as string[])
-      .concat(
-        path !== undefined
-          ? typeof path == 'string'
-            ? [path]
-            : [...path]
-          : [],
-        this.DEFAULT_CONFIG_FILES,
-      )
+      .concat(path || [], this.DEFAULT_CONFIG_FILES)
       .filter(
         (file) =>
           fs.existsSync(file) && ConfigurationParserFactory.supports(file),

--- a/src/configuration/configuration.registry.ts
+++ b/src/configuration/configuration.registry.ts
@@ -11,7 +11,7 @@ import {
  * with Configuration decorator.
  */
 export class ConfigurationRegistry {
-  private static readonly registry: any[] = [];
+  private static readonly registry: Array<unknown> = [];
 
   /**
    * Registers a type.

--- a/src/configuration/configuration.registry.ts
+++ b/src/configuration/configuration.registry.ts
@@ -11,7 +11,7 @@ import {
  * with Configuration decorator.
  */
 export class ConfigurationRegistry {
-  private static readonly registry = [];
+  private static readonly registry: any[] = [];
 
   /**
    * Registers a type.

--- a/src/configuration/parsers/configuration-parser.factory.ts
+++ b/src/configuration/parsers/configuration-parser.factory.ts
@@ -26,7 +26,7 @@ export class ConfigurationParserFactory {
    * @returns {ConfigurationParser}      the configuration parser
    */
   static getParser(file: string): ConfigurationParser {
-    const ext = this.getFileExt(file);
+    const ext = this.getFileExt(file) as keyof typeof this.parsers;
     return this.parsers[ext];
   }
 
@@ -48,6 +48,6 @@ export class ConfigurationParserFactory {
    * @returns {string}      the file extension
    */
   private static getFileExt(file: string): string {
-    return file.split('.').pop();
+    return file.split('.').pop() || '';
   }
 }

--- a/src/configuration/parsers/configuration-parser.factory.ts
+++ b/src/configuration/parsers/configuration-parser.factory.ts
@@ -12,7 +12,7 @@ export class ConfigurationParserFactory {
   /**
    * The supported file configuration parsers
    */
-  private static readonly parsers = {
+  private static readonly parsers: Record<string, ConfigurationParser> = {
     env: new DotEnvConfigurationParser(),
     yml: new YamlConfigurationParser(),
     yaml: new YamlConfigurationParser(),
@@ -38,7 +38,7 @@ export class ConfigurationParserFactory {
    */
   static supports(file: string): boolean {
     const ext = this.getFileExt(file);
-    return this.parsers.hasOwnProperty(ext);
+    return ext ? this.parsers.hasOwnProperty(ext) : false;
   }
 
   /**
@@ -47,7 +47,7 @@ export class ConfigurationParserFactory {
    * @param   {string} file the file name
    * @returns {string}      the file extension
    */
-  private static getFileExt(file: string): string {
-    return file.split('.').pop() || '';
+  private static getFileExt(file: string): string | undefined {
+    return file.split('.').pop();
   }
 }

--- a/src/configuration/parsers/yaml-configuration.parser.ts
+++ b/src/configuration/parsers/yaml-configuration.parser.ts
@@ -13,6 +13,6 @@ export class YamlConfigurationParser implements ConfigurationParser {
    * @returns {Record<string, any>}      an object representation of the configuration file
    */
   public parse(file: string): Record<string, any> {
-    return yaml.load(fs.readFileSync(file, 'utf-8'));
+    return yaml.load(fs.readFileSync(file, 'utf-8')) as Record<string, any>;
   }
 }

--- a/src/configuration/resolvers/aws/parameter-store-configuration.resolver.ts
+++ b/src/configuration/resolvers/aws/parameter-store-configuration.resolver.ts
@@ -36,7 +36,7 @@ export class AwsParameterStoreConfigurationResolver
     if (errors && errors.length) {
       throw new Error(
         `Unable to resolve parameter:\n${errors
-          .map((e) => `${e.key}: ${e.id} - ${e.error.message}`)
+          .map((e) => `${e.key}: ${e.id} - ${e.error?.message}`)
           .join('\n')}`,
       );
     }

--- a/src/configuration/resolvers/aws/secrets-manager-configuration.resolver.ts
+++ b/src/configuration/resolvers/aws/secrets-manager-configuration.resolver.ts
@@ -39,7 +39,7 @@ export class AwsSecretsManagerConfigurationResolver
     if (errors && errors.length) {
       throw new Error(
         `Unable to resolve secrets:\n${errors
-          .map((e) => `${e.key}: ${e.id} - ${e.error.message}`)
+          .map((e) => `${e.key}: ${e.id} - ${e.error?.message}`)
           .join('\n')}`,
       );
     }

--- a/src/interpolation/variables.ts
+++ b/src/interpolation/variables.ts
@@ -37,7 +37,7 @@ export class Variables {
    * @returns {Record<string, unknown>}        The expanded object
    */
   static expand(record: Record<string, unknown>): Record<string, unknown> {
-    const expanded = {};
+    const expanded: Record<string, unknown> = {};
     for (const key in record) {
       const value = record[key];
       const interpolated = this.interpolate(value, record);

--- a/test/configfy.module.spec.ts
+++ b/test/configfy.module.spec.ts
@@ -34,7 +34,7 @@ describe('ConfigifyModule', () => {
         configFilePath: file,
       });
 
-      const provider = module.providers.filter(
+      const provider = module.providers?.filter(
         (p) =>
           (p as ValueProvider).useValue.constructor.name ===
           ComplexDotEnvConfiguration.name,
@@ -66,7 +66,7 @@ describe('ConfigifyModule', () => {
         configFilePath: file,
       });
 
-      const provider = module.providers.filter(
+      const provider = module.providers?.filter(
         (p) =>
           (p as ValueProvider).useValue.constructor.name ===
           ComplexYmlConfiguration.name,
@@ -98,7 +98,7 @@ describe('ConfigifyModule', () => {
         configFilePath: file,
       });
 
-      const provider = module.providers.filter(
+      const provider = module.providers?.filter(
         (p) =>
           (p as ValueProvider).useValue.constructor.name ===
           ComplexJsonConfiguration.name,

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -12,10 +12,10 @@
     "baseUrl": "./",
     "incremental": true,
     "skipLibCheck": true,
-    "strictNullChecks": false,
-    "noImplicitAny": false,
-    "strictBindCallApply": false,
-    "forceConsistentCasingInFileNames": false,
-    "noFallthroughCasesInSwitch": false
+    "strictNullChecks": true,
+    "noImplicitAny": true,
+    "strictBindCallApply": true,
+    "forceConsistentCasingInFileNames": true,
+    "noFallthroughCasesInSwitch": true
   }
 }

